pkg/upstream/cluster/loadbalancer.go

```diff
@@ -624,3 +626,167 @@ func (lb *reqRoundRobinLoadBalancer) IsExistsHosts(metadata api.MetadataMatchCri
+			return host
+		}
+		return nil
+	}
+
+	var candidate types.Host
+
+	// If `total` is less than or equal to `choice`, we can iterate over all elements directly.
+	if total <= int(lb.choice) {
+		candidate = lb.iterateChoose()
+	} else {
+		candidate = lb.randomChoose()
+		if candidate == nil {
+			if log.DefaultLogger.GetLogLevel() >= log.WARN {
+				log.DefaultLogger.Warnf("[lb][shortest_response] no host chosen after %d choice, fallback to WRR", lb.choice)
+			}
+			return lb.wrrLoadBalancer.ChooseHost(context)
+		}
+	}
+
+	return candidate
+}
+
+func (lb *shortestResponseLoadBalancer) IsExistsHosts(_ api.MetadataMatchCriteria) bool {
+	return lb.hosts.Size() > 0
+}
+
+func (lb *shortestResponseLoadBalancer) HostNum(_ api.MetadataMatchCriteria) int {
+	return lb.hosts.Size()
+}
+
+func (lb *shortestResponseLoadBalancer) iterateChoose() types.Host {
+	var candidate types.Host
+	var candidateScore float64
+
+	lb.hosts.Range(func(temp types.Host) bool {
+		if !temp.Health() {
+			return true
+		}
+
+		tempScore := shortestResponseScore(temp)
+		if candidate == nil || tempScore < candidateScore {
+			candidate = temp
+			candidateScore = tempScore
+		}
+
+		return true
+	})
+
+	if candidate != nil {
+		if log.DefaultLogger.GetLogLevel() >= log.DEBUG {
+			log.DefaultLogger.Debugf("[lb][shortest_response] iterate choose host %s with score %d", candidate.AddressString(), candidateScore)
+		}
+	}
+	return candidate
+}
+
+// Choose `choice` times and return the best one.
+// See The Power of Two Random Choices: A Survey of Techniques and Results
+//
+//	http://www.eecs.harvard.edu/~michaelm/postscripts/handbook2001.pdf
+func (lb *shortestResponseLoadBalancer) randomChoose() types.Host {
+	total := lb.hosts.Size()
+
+	var candidate types.Host
+	var candidateScore float64
+
+	lastIdx := 0
+
+	for choice := int(lb.choice); choice > 0; choice-- {
+		n := total - lastIdx - choice + 1
+
+		lb.mutex.Lock()
+		idx := lb.rand.Intn(n) + lastIdx
+		lb.mutex.Unlock()
+
+		lastIdx = idx
+
+		temp := lb.hosts.Get(idx)
+		if !temp.Health() {
+			continue
+		}
+
+		tempScore := shortestResponseScore(temp)
+		if candidate == nil || tempScore < candidateScore {
+			candidate = temp
+			candidateScore = tempScore
+		}
+	}
+
+	if candidate != nil {
+		if log.DefaultLogger.GetLogLevel() >= log.DEBUG {
+			log.DefaultLogger.Debugf("[lb][shortest_response] random choose host %s with score %f", candidate.AddressString(), candidateScore)
+		}
+	}
+	return candidate
+}
+
+func shortestResponseScore(h types.Host) float64 {
+	stats := h.HostStats()
+
+	responseSuccessRate := 1 - stats.UpstreamResponseFailedEWMA.Rate()
+	if responseSuccessRate == 0 {
+		responseSuccessRate = 1e-6
+	}
+
+	duration := stats.UpstreamRequestDurationEWMA.Rate()
+	if duration == 0 {
+		duration = 1e-6
+	}
+
+	// All factors cannot be 0 to ensure that other factors can take effect normally
+	return duration * float64(stats.UpstreamRequestActive.Count()+1) / float64(h.Weight()) / responseSuccessRate
```

Comment:
One question, why do we need to multiply the duration by the number of requests?
As the duration is the result of the number of requests. For a given node, a high number of requests usually mean a higher duration. Multiplying these two factors seems to amplify the difference. The ingress-nginx's ewma implementation only considers the duration.

Multiplying the number of requests has a side effect - now the shortest load balancer algorithm is not self-balancing. Considering two nodes (1C and 2C) have the same duration, they usually have different numbers of requests, for example, the 2C node may have a double number of requests than the 1C. Therefore, we need to configure a weight for the 2C node, otherwise they are not balanced. However, if we remove the number of requests, we don't need to introduce weights for different nodes.

Reply:
In the real world, there is a singularity in the performance curve of the server with the number of requests processed at the same time (usually related to the number of CPU cores or other hardware resources). Before reaching this singularity, the performance is very good, but when it reaches this A sudden change in server performance occurs during the singularity.

To give a practical scenario, in a CPU-intensive service, before the number of CPU cores is less than the number of CPU cores, the service always performs well and has a uniform performance, but it will change when the number of concurrently processed requests exceeds the number of CPU cores. Either the frequent context switching causes the delay of all requests to increase, or it enters the queue for queuing, so as the queue gets longer and longer, the waiting time becomes longer and longer.

Therefore, although the performance of this server is very good, we still need to consider their load situation to avoid becoming the last straw of the compressed camel.

Reply:
Linkerd also uses the EWMA algorithm weighted by the number of connections. This part of the logic is not in linkerd itself, but in tower-balance

See: https://github.com/linkerd/linkerd2/blob/cea37ee48d7cc93b5035598bf9b4da0db5a9e300/proxy/src/outbound.rs#L86-L89
and 
https://github.com/tower-rs/tower/blob/0ab48a2e6db55fabaa95370f304fbc0caf0616a6/tower-load/src/peak_ewma.rs#L175-L190

Reply:
I can agree with some of your concerns.
However, the current implementation focuses more on the number of connections than the response duration.

There are two ways that the number of connections can affect the final result:
1. the larger number of connections, the longer duration
2. the score is multiplied by the number of connections

while the response duration only has a way to affect the final result.

> From another point of view, EWMA can also be considered as the weighting of the number of active requests, which is used to predict which server will reach the empty state faster (there is such an informal statement)

Yes, it is exactly what the current implementation does.

I would recommend renaming this loadbalancer to `LeastRequestWithEWMA`, so its name can reflect what it is. In fact, this loadbalancer will select a node with the least request instead of the shortest duration. For example,

Considering we have two nodes, node A has 2C, and node B has 1C. The number of active requests on A is 3, and 2 on B. As A has 2C, the estimated response duration of A is smaller than B (3/2 < 2/1).
For a loadbalancer which will always select the node with the shortest response, A will be chosen.
However, for the loadbalancer in this PR,
A has score 4.5, and B has score 4. B will be chosen because it has a lower score, even it has a higher estimated response duration.

Maybe a new name can solve the conflict and reach an agreement.

Reply:
I see what you mean, I will try to implement a lock-free real-time EWMA, so that we can no longer use the number of connections and weights, but simply load balance based on response time.

Reply:
I think I understand what you mean. Indeed, this is a problem especially when the difference in response time (std.) is orders of magnitude different from the response time itself, then changes in the number of connections will be more sensitive than the response time. But on the other hand, we can't just rely on the response time. As mentioned earlier, many times the increase in the response time is a jump. It is possible to use Round Robin without encountering a singularity, but because the EWMA will not be updated before the request returns. , then a large number of requests may flood into the optimal node at the same time, leading to the singularity.

I will still keep the number of connections factor, but I will try to reduce its impact, the logarithm may be a good means

Reply:
> But on the other hand, we can't just rely on the response time.

Totally agree.

> I will still keep the number of connections factor

It is fine to keep the number of connections. But I think we can find a better name for this loadbalancer. There are many factors to select a node in it, the shortest response is just one of them. And this loadbalancer doesn't guarantee the node with the shortest response is chosen, even if the weights are equal. Unlike the least request loadbalancer, which will choose the node with the least active requests. I think this name is misleading and causes many conflicts. Maybe we can rename it, for instance, calling it EWMA or something else.

Reply:
Hello, @muyuan0 . The name of the load balancing algorithm has been changed. 

And you are right, it has been verified that the weights cannot simply be factored into the calculation, and the cost of allocation according to the flow is very high (more than the benefit), so I try to use this as a hint to the number of requests: usually, more weights mean more computing resources, so calculate the number of active requests per unit of computing resources in the upstream through the weights, which is the same as the number of weighted active requests. 

Thank you very much for your advice and patience explanation of.